### PR TITLE
Feature/testing

### PR DESCRIPTION
**Pull request by Geist5000**
added cypress testing to the project

### CODE REVIEW

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+Testaufwand enorm weil diese komplett automatisch abläuft.
+
+Damit Cypress korrekt arbeiten kann müssen wir ersteinmal sicherstellen das Cypress alle nötigen Module von Node hat.
+
+## Module-Dependencys
+Module: "smtp-tester"
+Beschreibung: 
+Wird benötigt damit Cypress einen smtp-server für seine Testumgebung erstellt.
+```shell
+npm i smtp-tester
```
Comment (Geist5000) [inline TESTING.md:12]: is embedded into package.json. Its enough to reference `npm install command`

File: cypress.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  "env": {
+    "DB_HOST": "127.0.0.1",
+    "DB_USER": "root",
+    "DB_PASSWORD": "",
+    "DB_PORT": 3306,
+    "BACKEND_URL": "http://localhost/toolbox-backend/public/",
+    "APP_CLIENT_ID": 2,
+    "TEST_LOGIN": {
+      "username": "max@test.test",
+      "password": "password"
```
Comment (Geist5000) [inline cypress.json:21]: Put this into two env variables, a.e. "TEST_LOGIN_USERNAME, TEST_LOGIN_PASSWORD"

File: cypress/fixtures/user.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "email": "max@test.test",
+  "password": "password"
+}
```
Comment (Geist5000) [inline cypress/fixtures/user.json:4]: remove this or testLoginData.json

File: cypress/integration/tool_tests/swot-analysis_I.js [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    }else if (key == "desc")
+    {
+
+      cy.wrap($input).should('have.value', swot[result[1]][result[2]]["desc"])
+      
+    }
+    
+  }
+}
+function CreateSave()
```
Comment (Geist5000) [inline cypress/integration/tool_tests/swot-analysis_I.js:235]: Move to command with arguments (name, data, tool_id, owner and description as command)

File: cypress/integration/tool_tests/swot-analysis_I.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+          })
+
+          cy.log("Additional Cards added and filled")
+
+          //
+          cy.contains("Speichern")
+          .click();
+          
+          //TODO: This POST gives back a NULL is that supposed to be so?
```
Comment (Geist5000) [inline cypress/integration/tool_tests/swot-analysis_I.js:166]: Resolve TODO or remove TODO

File: cypress/integration/validation/profile.js [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        .click()
+        cy.wait("@user")
+        .its("response")
+        .should('include',{
+            statusCode: 401
+          })
+
+          
+        cy.intercept("GET",/.*settings.*/).as('get')
+        cy.reload();
```
Comment (Geist5000) [inline cypress/integration/validation/profile.js:51]: check if neccessary

File: cypress/integration/validation/profile.js [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        cy.get('@get')
+        .its("response")
+        .should("include",{
+            statusCode: 200
+        })
+        
+        cy.get("input[id='username']")
+        .should('have.value',"test_user")
+
+        //REVERT BACK to original
```
Comment (Geist5000) [inline cypress/integration/validation/profile.js:62]: Move to start of test

File: cypress/integration/validation/profile.js [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+          })
+
+        cy.get('div [class="feedback text-success"]')
+        .should('not.be.visible')
+
+        cy.reload();
+        cy.get("input[id='email']")
+        .should('have.value',"max@test.test")
+
+        //REVERT BACK to original
```
Comment (Geist5000) [inline cypress/integration/validation/profile.js:104]: see previous

File: cypress/integration/validation/profile.js [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+        cy.get('div [class="feedback text-success"]')
+        .should('not.be.visible')
+
+        cy.reload();
+        cy.get("input[id='email']")
+        .should("max@test.test")
+
+
+        //REVERT BACK to original
```
Comment (Geist5000) [inline cypress/integration/validation/profile.js:148]: see previous

File: cypress/integration/validation/registration.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        .its("response.body")
+        .should('nested.include',{"data.available": true
+        })
+        cy.get("div")
+        .contains("E-Mail ist Verfügbar!")
+        .should("be.visible")
+
+        //--
+        cy.get('#username').type('Tim_Tester')
+        cy.intercept('GET', /.*api\/checkUsername.*/).as('checkUser')
```
Comment (Geist5000) [inline cypress/integration/validation/registration.js:21]: same same

File: cypress/plugins/index.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    // `config` is the resolved Cypress config
+
+    // Usage: cy.task('queryDb', query)
+
+
+    if (config.testingType === 'component') {
+        require('@cypress/react/plugins/react-scripts')(on, config)
+    } else {
+        // starts the SMTP server at localhost:7777
+        const port = 7777;
```
Comment (Geist5000) [inline cypress/plugins/index.js:30]: put port setting into an env variable

**Review by Geist5000 — COMMENTED** [review]
please fix comments

File: cypress/integration/tool_tests/swot-analysis_II.js [hunk 12]
```diff
@@ -3,7 +3,7 @@ const { wait } = require("@testing-library/react");
 describe('SWOT Analyisis Part II', () => {
     beforeEach(() =>{
         cy.task("queryDb",`DELETE FROM toolbox.saves WHERE owner_id= 1 AND name= "TEST-SWOT VON MAX";`);
-        CreateSave()
+        cy.CreateSave("swot-1","TEST-SWOT VON MAX",2)
```
Comment (Geist5000) [inline cypress/integration/tool_tests/swot-analysis_II.js:6]: shouldn't this be swot-2?

File: cypress/integration/validation/profile.js [hunk 13]
```diff
@@ -2,6 +2,11 @@ const { debug } = require("console")
 
 describe('Checking Profile', () => {
     beforeEach(() =>{
+        cy.task("queryDb",
+        `UPDATE toolbox.users
+        SET username = "test_user", email = "max@test.test", password = "$2y$10$UJinKiSroTM2MsFmLBu2puwOJynkzmTj9ZgLnsOHfa3HIK43/QKxW"
+        WHERE id = 1`)
+
         cy.loginViaApi("max@test.test", "password")
```
Comment (Geist5000) [inline cypress/integration/validation/profile.js:10]: shouldn't this login data retrieved by env variables?  (loginViaApi and in update statement

File: cypress/support/commands.js [hunk 14]
```diff
@@ -85,3 +88,28 @@ Cypress.Commands.add("visitSite",(site ,clickOn)=>{
...
 })
+
+//Parameter
+//saveSlot string: Expample: "swot-1" for a first step save for SWOT, which slot has what save look into saveDummyData.json in fixtures
+//name string: name for the save
+//tool_id int : Which tool, 2 for SWOT-Analysis
+//owner_id int: optional: standard is 0 for max@test.test
+Cypress.Commands.add('CreateSave',(saveSlot,name,tool_id,owner_id = 1)=>{
+    
+    cy.fixture("saveDummyData.json").then(function (SaveData)
```
Comment (Geist5000) [inline cypress/support/commands.js:99]: the dummy data is currently forced to be in one file for **all** tests. This is could get very messy very fast. I would suggest to move every save data in its own file.

I would suggest to create two files, one named `swot-1.json` and one `swot-2.json`.

Comment (Geist5000) [reply cypress/support/commands.js:99]: Maybe put these files inside its own folder inside the fixtures folder

**Review by Geist5000 — COMMENTED** [review]
pleas look at my new comments

File: TESTING.md (hunk 1 above)
Comment (Geist5000) [reply TESTING.md:12]: Why did you resolve this?

File: cypress/integration/validation/registration.js (hunk 10 above)
Comment (Geist5000) [reply cypress/integration/validation/registration.js:21]: why did you resolve this?

File: cypress/integration/validation/profile.js (hunk 6 above)
Comment (Geist5000) [reply cypress/integration/validation/profile.js:56]: whats your result of checking if this is neccessary? Please comment your results

**Comment by Geist5000** [comment]
Please look at the comments of the old review, I unresolved some of them

File: cypress/integration/validation/profile.js (hunk 6 above)
Comment (Maruun86) [reply cypress/integration/validation/profile.js:56]: It's for the purpse to get the former values back in that moment.

` //username
        cy.get("input[id='username']")
        .clear()
        .type("test_user1")

        cy.get("input[id='current_password']")
        .clear()
        .type("passwort") //invalid

        cy.intercept("POST", /.*api\/users.*/).as("user")
        cy.contains("Änderungen speichern")
        .click()
        cy.wait("@user")
        .its("response")
        .should('include',{
            statusCode: 401
          })

          
        cy.intercept("GET",/.*settings.*/).as('get')
        cy.reload();
        cy.wait("@get")
        cy.get('@get')
        .its("response")
        .should("include",{
            statusCode: 200
        })
    cy.get("input[id='username']")
        .should('have.value',"test_user")
`

In this case cypress try's to change the username, it should fail, i reload the page and check if the username has changed.
It also reset's the form to its true values.

**Review by Geist5000 — COMMENTED** [review]
nearly done. Looks good for the most part

**Review by Geist5000 — COMMENTED** [review]
looks good

**Pull request merged**